### PR TITLE
Feature: Canvas Layout Options(max width config)

### LINKED ISSUE #2577: [Feature] Allow users to configure the max / min stretch width of an app
## Summary

Today our canvas is at a fixed width and on different devices, this causes either a large white space or a horizontal scroll to appear. Allowing users to configure a range within which they would like the app to stretch would allow them to design the optimal experience for the devices they want to support.

## Motivation

A user had reached out with a use case of using appsmith on a tablet. They did not need a fully responsive UI but simply a way to design the UI so that it would fit correctly on the screen size they needed. This is a simple solution and can be the first step towards becoming fully responsive. The extension to this could be to allow users to specify breakpoints that will trigger reflow.

### PR DESCRIPTION
**Pull request by marks0351**
> Pull Request Template
>
> Use this template to quickly create a well written pull request. Delete all quotes before creating the pull request.

## Description
 Adding control to select the layout(Default, Tablet, Mobile, Fluid) of the application

Fixes #2577 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How Has This Been Tested?

> Please describe the tests that you ran to verify your changes. Provide instructions, so we can reproduce. 
> Please also list any relevant details for your test configuration.

- Test A
- Test B

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/pages/Editor/Explorer/hooks.ts [hunk 1]
```diff
@@ -236,3 +246,58 @@ export const useEntityEditState = (entityId: string) => {
     (state: AppState) => state.ui.explorer.editingEntityName === entityId,
   );
 };
+
+export const useDynamicAppLayout = () => {
```
Comment (riodeuno) [inline app/client/src/pages/Editor/Explorer/hooks.ts:250]: Let's move the hooks to the `utils/hooks` folder.

File: app/client/src/pages/Editor/MainContainerLayoutControl.tsx [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  align-items: center;
+  .dropdown-wrapper {
+    margin-left: 220px;
+    width: 260px;
+  }
+  .bp3-popover-target {
+    pointer-events: none;
+  }
+  .layout-control {
+    pointer-events: all;
```
Comment (riodeuno) [inline app/client/src/pages/Editor/MainContainerLayoutControl.tsx:76]: Please figure out if we can avoid updating the styles of our components, as much as possible.

File: app/client/src/reducers/entityReducers/canvasWidgetsReducer.tsx [hunk 3]
```diff
@@ -5,7 +5,11 @@ import {
   ReduxAction,
 } from "constants/ReduxActionConstants";
 import { WidgetProps } from "widgets/BaseWidget";
-import { UpdateWidgetPropertyPayload } from "actions/controlActions";
+import {
+  UpdateCanvasLayout,
+  UpdateWidgetPropertyPayload,
+} from "actions/controlActions";
+import defaultTemplate from "templates/default";
```
Comment (riodeuno) [inline app/client/src/reducers/entityReducers/canvasWidgetsReducer.tsx:12]: Let's use `MAIN_CONTAINER_WIDGET_ID` from `WidgetConstants`

File: app/client/src/reducers/entityReducers/pageListReducer.tsx [hunk 4]
```diff
@@ -105,11 +105,18 @@ export const pageListReducer = createReducer(initialState, {
   },
 });
 
+type AppLayoutType = "FIXED" | "FLUID";
```
Comment (riodeuno) [inline app/client/src/reducers/entityReducers/pageListReducer.tsx:108]: Is this the same as ContainerLayoutType?

File: app/client/src/sagas/ApplicationSagas.tsx [hunk 5]
```diff
@@ -217,6 +217,25 @@ export function* setDefaultApplicationPageSaga(
   }
 }
 
+function* updateApplicationLayout(
```
Comment (riodeuno) [inline app/client/src/sagas/ApplicationSagas.tsx:220]: Let's rename this to updateApplicationLayoutSaga

**Review by riodeuno — CHANGES_REQUESTED** [review]

**Review by hetunandu — APPROVED** [review]

**Review by hetunandu — APPROVED** [review]

**Review by hetunandu — APPROVED** [review]

File: app/client/src/utils/hooks/useDynamicAppLayout.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import { AppState } from "reducers";
+import { getWidget } from "sagas/selectors";
+import { getAppMode } from "selectors/applicationSelectors";
+import {
+  getCurrentApplicationLayout,
+  getCurrentPageId,
+} from "selectors/editorSelectors";
+import { useWindowSizeHooks } from "./dragResizeHooks";
+
+export const useDynamicAppLayout = () => {
```
Comment (riodeuno) [inline app/client/src/utils/hooks/useDynamicAppLayout.tsx:16]: Let's talk about how we can mock redux and sagas to write unit tests for these types of functions.

**Review by riodeuno — APPROVED** [review]
Looks good to me!

**Pull request merged**